q2cli/commands.py

```diff
@@ -0,0 +1,169 @@
+
+    def get_command(self, ctx, name):
+        if name in self._builtin_commands:
+            return self._builtin_commands[name]
+
+        try:
+            plugin = self._plugin_lookup[name]
+        except KeyError:
+            return None
+
+        # TODO: pass short_help=plugin.description, pending its
+        # existence: https://github.com/qiime2/qiime2/issues/81
+        support = 'Getting user support: %s' % plugin.user_support_text
+        citing = 'Citing this plugin: %s' % plugin.citation_text
+        website = 'Plugin website: %s' % plugin.website
+        help_ = '\n\n'.join([website, support, citing])
+
+        return PluginCommand(plugin, ctx, short_help='', help=help_)
+
+
+class PluginCommand(click.MultiCommand):
+    def __init__(self, plugin, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # the cli currently doesn't differentiate between methods
+        # and visualizers.
+        self._plugin = plugin
+        self._action_lookup = {q2cli.util.normalize_name(k): v for k, v in
+                               itertools.chain(plugin.methods.items(),
+                                               plugin.visualizers.items())}
+
+    def list_commands(self, ctx):
+        return sorted(self._action_lookup)
+
+    def get_command(self, ctx, name):
+        try:
+            action = self._action_lookup[name]
+        except KeyError:
+            click.echo("Error: QIIME plugin %r has no action %r."
+                       % (self._plugin.name, name), err=True)
+            ctx.exit(2)  # Match exit code of `return None`
+
+        if type(action) is qiime.sdk.Method:
+            extension = qiime.sdk.Artifact.extension
+        else:
+            extension = qiime.sdk.Visualization.extension
+
+        return ActionCommand(name, action, extension)
+
+
+class ActionCommand(click.Command):
+    def __init__(self, name, action, extension):
+        self.name = name
+        self.action = action
+        self.extension = extension
+        self.generated_handlers = self.build_generated_handlers()
+        self.output_dir_handler = q2cli.handlers.OutputDirHandler()
+
+        super().__init__(name, params=list(self.get_click_parameters()),
+                         callback=self, short_help=action.name,
+                         help=action.description)
+
+    def build_generated_handlers(self):
+        handlers = collections.OrderedDict()
+        handler_map = collections.OrderedDict([
+            ('inputs', q2cli.handlers.ArtifactHandler),
+            ('parameters', q2cli.handlers.parameter_handler_factory),
+            ('outputs', functools.partial(q2cli.handlers.ResultHandler,
+                                          self.extension))
+        ])
+
+        for group_type, constructor in handler_map.items():
+            grp = getattr(self.action.signature, group_type)
+            for name, (semtype, _) in sorted(grp.items(), key=lambda x: x[0]):
```

Comment:
Add back comment indicating to remove sorted when `Signature` preserves API order.